lib/Conversion/TorchToLinalg/Pooling.cpp

```diff
@@ -861,67 +699,82 @@ class ConvertAtenAdaptiveMaxPool2dOp
     Value kIter = rewriter.create<tensor::EmptyOp>(
         loc, getAsOpFoldResult(kIterSizeVector), boolType);
 
-    // need to buffer input, else there will possibly be an out of bounds access
-    // later buffVal = 0 for avg pooling and -inf for max pooling
-    auto smallestFPValueAttr = rewriter.getFloatAttr(
-        elementType,
-        APFloat::getInf(elementType.cast<mlir::FloatType>().getFloatSemantics(),
-                        /*Negative=*/true));
-    Value buffVal = rewriter.create<arith::ConstantOp>(loc, elementType,
-                                                       smallestFPValueAttr);
-    SmallVector<int64_t> lowPadding(rank, 0);
-    SmallVector<int64_t> highPadding(2, 0);
-    for (int i = 0; i < rank - 2; i++) {
-      highPadding.push_back(1);
-    }
-    Value buffInput = torch_to_linalg::getPaddedTensor(
-        op, rewriter, input, lowPadding, highPadding, buffVal);
-
-    // make a list of outputSizes
+    // get output sizes used for initializing some tensors
     SmallVector<Value> outputSizes;
-    for (unsigned i = 0; i < 2; i++) {
+    for (unsigned i = 0; i < nonSpatial; i++) {
       outputSizes.push_back(getDimOp(rewriter, loc, input, i));
     }
-    for (unsigned i = 2; i < rank; i++) {
-      outputSizes.push_back(outShapeIndexVector[i - 2]);
+    for (unsigned i = 0; i < rank - nonSpatial; i++) {
+      outputSizes.push_back(outShapeIndexVector[i]);
     }
 
-    // for avg pooling the auxTensor should hold kernel widths (kSizeTensor)
-    // for max Pooling, it should hold the indices
-    RankedTensorType outputType1 =
-        typeConverter->convertType(op.getResult1().getType())
-            .cast<RankedTensorType>();
-    Type indicesType = outputType1.getElementType();
-    Value auxTensor = rewriter.create<tensor::EmptyOp>(
-        loc, getAsOpFoldResult(outputSizes), indicesType);
+    // get outputType and initialize an auxTensor
+    RankedTensorType outputType, auxTensorType;
+    Value buffVal, auxTensor;
+    if constexpr (isMaxPool) {
+      outputType = typeConverter->convertType(op.getResult0().getType())
+                       .template cast<RankedTensorType>();
+      auxTensorType = typeConverter->convertType(op.getResult1().getType())
+                          .template cast<RankedTensorType>();
+      Type auxTensorElementType = auxTensorType.getElementType();
+      auto smallestFPValueAttr = rewriter.getFloatAttr(
+          elementType,
+          APFloat::getInf(
+              elementType.cast<mlir::FloatType>().getFloatSemantics(),
+              /*Negative=*/true));
+      buffVal = rewriter.create<arith::ConstantOp>(loc, elementType,
+                                                   smallestFPValueAttr);
+      auxTensor = rewriter.create<tensor::EmptyOp>(
+          loc, getAsOpFoldResult(outputSizes), auxTensorElementType);
+    } else {
```

Comment:
I would have

else if (isAvgPool) { ... }

else {
  unreachable error of some sort 
}

Reply:
Likewise for the other 2 places where there is a branching between the reduction types. 

Did you consider using old-fashioned polymorphism with virtual functions instead of if-else approach? It might be neater  (This isn't a review request, I'm just curious!).

Reply:
> I would have
> 
> else if (isAvgPool) { ... }
> 
> else { unreachable error of some sort }

Is there a benefit to this change beyond readability? I can't imagine that the compiler would get past the static_assert at the beginning.

> Did you consider using old-fashioned polymorphism with virtual functions instead of if-else approach? It might be neater (This isn't a review request, I'm just curious!).

Hm, I hadn't considered that. Originally, I was offloading pieces to helper functions (similar to createPoolingOp in this file), but realized that I either had to duplicate a bunch of boilerplate or pass a lot of variables to each function, so I opted to just write a few if statements instead.  

Although, I'm thinking now that it might be worth the bit of redundancy to do something more modular, since it may be inconvenient to edit this pass in the future if one of these pytorch functions gets modified. For example, if adaptive max pooling with and without indices are separated, then someone would likely need to read through the whole pass to provide a fix for that.

I don't think it would take too long to convert the if statement cases into polymorphic helper functions, so I'm happy to try out this approach if you want.